stacks-common/src/util/lru_cache.rs

```diff
@@ -0,0 +1,596 @@
+                    Ok(()) => node.dirty = false,
+                    Err(e) => return Ok(Err(e)),
+                }
+                node.dirty = false;
+            }
+            current = next;
+        }
+        Ok(Ok(()))
+    }
+
+    /// Helper function to remove a node from the linked list (by index)
+    fn detach_node(&mut self, index: usize) -> Result<(), LruCacheCorrupted> {
+        let node = self.order.get(index).ok_or(LruCacheCorrupted)?;
+        let prev = node.prev;
+        let next = node.next;
+
+        if index == self.tail {
+            // If this is the last node, update the tail to point to its previous node
+            self.tail = prev;
+        } else {
+            // Else, update the next node to point to the previous node
+            let next_node = self.order.get_mut(next).ok_or(LruCacheCorrupted)?;
+            next_node.prev = prev;
+        }
+
+        if index == self.head {
+            // If this is the first node, update the head to point to the next node
+            self.head = next;
+        } else {
+            // Else, update the previous node to point to the next node
+            let prev_node = self.order.get_mut(prev).ok_or(LruCacheCorrupted)?;
+            prev_node.next = next;
+        }
+
+        Ok(())
+    }
+
+    /// Helper function to attach a node as the head of the linked list
+    fn attach_as_head(&mut self, index: usize) -> Result<(), LruCacheCorrupted> {
+        let node = self.order.get_mut(index).ok_or(LruCacheCorrupted)?;
+        node.prev = self.capacity;
+        node.next = self.head;
+
+        if self.head != self.capacity {
+            // If there is a head, update its previous pointer to this one
+            let head_node = self.order.get_mut(self.head).ok_or(LruCacheCorrupted)?;
+            head_node.prev = index;
+        } else {
+            // Else, the list was empty, so update the tail
+            self.tail = index;
+        }
+        self.head = index;
+        Ok(())
+    }
+
+    /// Helper function to move a node to the head of the linked list
+    fn move_to_head(&mut self, index: usize) -> Result<(), LruCacheCorrupted> {
+        if index == self.head {
+            // If the node is already the head, do nothing
+            return Ok(());
+        }
+
+        self.detach_node(index)?;
+        self.attach_as_head(index)
+    }
+
+    /// Helper function to evict the least-recently used node, which is the
+    /// tail of the linked list
+    /// Returns the index of the evicted node
+    fn evict_lru(&mut self) -> Result<usize, LruCacheCorrupted> {
+        let index = self.tail;
+        if index == self.capacity {
+            // If the list is empty, do nothing
+            return Ok(self.capacity);
+        }
+        self.detach_node(index)?;
+        let node = self.order.get(index).ok_or(LruCacheCorrupted)?;
+        self.cache.remove(&node.key);
+        Ok(index)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can you add some tests with capacity = 1?

Reply:
Added in 1742f8fce2eb54f9b7fafb6dde5ac514ebec8569.